src/main/java/org/commcare/xml/bulk/BulkElementParser.java

```diff
@@ -0,0 +1,173 @@
+import org.javarosa.xml.util.UnfullfilledRequirementsException;
+import org.kxml2.io.KXmlParser;
+import org.xmlpull.v1.XmlPullParserException;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.SortedMap;
+import java.util.TreeMap;
+
+/**
+ * A bulk element parser reads multiple types of the same transaction together into a buffer of
+ * TreeElements before performing a bulk processing step.
+ *
+ * NOTE: This parser will currently lose debugging information v. using a streamed processor,
+ * since the XML locations of potential errors are buffered.
+ *
+ * An implementing class should organize its implementation into the following steps:
+ *
+ * preParseValidate - Validate only the root element of the streaming xml doc, shouldn't advance
+ * the processing
+ *
+ * requestModelReadsForElement - Gather id's for models that will be needed to process a transaction
+ *
+ * performBulkRead - Read any relevant data from storage all at once
+ *
+ * processBufferedElement - After the bulk read, process the transactions one by one
+ *
+ * performBulkWrite - Write the processed models to storage
+ *
+ * Created by ctsims on 3/14/2017.
+ */
+
+public abstract class BulkElementParser<T> extends TransactionParser<TreeElement> {
+
+    protected int bulkTrigger = 500;
+    private List<TreeElement> currentBulkElementBacklog = new ArrayList<>();
+    private Set<String> currentBulkReadSet = new HashSet<>();
+    private Map<String, T> currentOperatingSet = new HashMap<>();
+    private LinkedHashMap<String, T> writeLog = new LinkedHashMap<>();
+
+    private int currentBulkReadCount = 0;
+
+    public BulkElementParser(KXmlParser parser) {
+        super(parser);
+    }
+
+    /**
+     * Sets the count at which the parser should trigger processing the current buffer
+     * @param newTriggerCount
+     */
+    protected void setBulkProcessTrigger(int newTriggerCount) {
+        this.bulkTrigger = newTriggerCount;
+    }
+
+    @Override
+    public TreeElement parse() throws InvalidStructureException, IOException,
+            XmlPullParserException, UnfullfilledRequirementsException {
+
+        preParseValidate();
+        TreeElement subElement = new TreeElementParser(parser, 0, "bulk_parser").parse();
+        requestModelReadsForElement(subElement, currentBulkReadSet);
+        currentBulkElementBacklog.add(subElement);
+        currentBulkReadCount++;
+
+        if (currentBulkReadCount > bulkTrigger) {
+            processCurrentBuffer();
+        }
+        return subElement;
+    }
+
+    public void processCurrentBuffer() throws IOException, XmlPullParserException,
+            InvalidStructureException {
+        currentOperatingSet = new HashMap<>();
+        performBulkRead(currentBulkReadSet, currentOperatingSet);
```

Comment:
@ctsims Just noting this is here too in case it got missed when you reviewed comments